RecoLocalCalo/EcalRecAlgos/src/EcalRecHitBuilderKernels.cu

```diff
@@ -0,0 +1,738 @@
+      ::ecal::reco::StorageScalarType const* time_ee,   
+      ::ecal::reco::StorageScalarType const* chi2_eb,   
+      ::ecal::reco::StorageScalarType const* chi2_ee,   
+      uint32_t const* flags_eb,
+      uint32_t const* flags_ee,
+      // output
+      uint32_t *did,
+      ::ecal::reco::StorageScalarType* energy,   // in energy [GeV]  
+      ::ecal::reco::StorageScalarType* time,  
+      ::ecal::reco::StorageScalarType* chi2,  
+      uint32_t* flagBits,
+      uint32_t* extra,
+      // other
+      int const nchannels,
+      uint32_t const offsetForInput,
+      uint32_t const offsetForHashes                     
+    ) {
+      
+      
+      //       
+      //    NB: energy   "type_wrapper<reco::StorageScalarType, L>::type" most likely std::vector<float>
+      //       
+      
+      int ch = threadIdx.x + blockDim.x*blockIdx.x;
+      
+      if (ch < nchannels) {
+        
+        int const inputCh = ch >= offsetForInput
+        ? ch - offsetForInput
+        : ch;
+        
+        uint32_t const * didCh = ch >= offsetForInput
+        ? did_ee
+        : did_eb;
+        
+        // only two values, EB or EE
+        // AM : FIXME : why not using "isBarrel" ?    isBarrel ? adc2gev[0] : adc2gev[1]
+        float adc2gev_to_use = ch >= offsetForInput
+        ? adc2gev[1]  // ee
+        : adc2gev[0]; // eb
+        
+        
+        // first EB and then EE
+        
+        ::ecal::reco::StorageScalarType const* amplitude = ch >= offsetForInput
+        ? amplitude_ee
+        : amplitude_eb;
+        
+        ::ecal::reco::StorageScalarType const* time_in = ch >= offsetForInput
+        ? time_ee
+        : time_eb;
+        
+        ::ecal::reco::StorageScalarType const* chi2_in = ch >= offsetForInput
+        ? chi2_ee
+        : chi2_eb;
+        
+        uint32_t const* flags_in = ch >= offsetForInput
+        ? flags_ee
+        : flags_eb;
+        
+        // simple copy
+        did[ch] = didCh[inputCh];
+        
+        auto const did_to_use = DetId{didCh[inputCh]};
+        
+        auto const isBarrel = did_to_use.subdetId() == EcalBarrel;
+        auto const hashedId = isBarrel
+        ? ecal::reconstruction::hashedIndexEB(did_to_use.rawId())
+        : offsetForHashes + ecal::reconstruction::hashedIndexEE(did_to_use.rawId());
+        
+        float const intercalib_to_use = intercalib[hashedId];
+        
+        
+        // get laser coefficient
+        float lasercalib = 1.;
+        
+        //
+        // AM: ideas
+        //
+        //    One possibility is to create the map of laser corrections once on CPU
+        //    for all crystals and push them on GPU.
+        //    Then only if the LS is different, update the laser correction
+        //    The variation within a LS is not worth pursuing (<< 0.1% !!)
+        //    and below the precision we can claim on the laser corrections (right?).
+        //    This will save quite some time (also for the CPU version?)    
```

Comment:
do the corrections actually change within a lumisection ?
or rather, does the interpolation has a granularity smaller than the lumisection ?

Reply:
Current correction is based on "time". Namely, we have two measurements, one every 40 minutes, and we perform a linear interpolation among them.
Based on the code I would say that the correction is really time dependent, time of the event.
   
`float interpolatedLaserResponse =   p_i / apdpnref + float(tt - t_i)  * (p_f - p_i)   / (apdpnref * float(t_f - t_i));
`

One possibility (just an idea) to speed up is to calculate the correction once every LS. From physics point of view, this should have ~ no effect (24 seconds step function vs linear interpolation that extend 40 minutes), but it can be tested with the current available data.
Computationally wise, this could save some time.

To be discussed with who wrote in first place the interpolation code to see if there is something subtle I am missing (special cases).

Reply:
Sounds like a very good idea !
(I always assumed the corrections were LS-based and handled by the EventSetup... this shows how far one can be wrong with reasonable assumptions)